### PR TITLE
add replay camera

### PR DESCRIPTION
**Pull request by SleepyPandaBear**
* WIP replay camera

### CODE REVIEW

File: src/robothub/replay_camera.py [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+    @property
+    def inputConfig(self) -> dai.Node.Input:
+        return self._input_control
+
+    @property
+    def inputControl(self) -> dai.Node.Input:
+        return self._input_control
+
+    @property
+    def isp(self) -> dai.Node.Output:
```
Comment (PetrNovota) [inline src/robothub/replay_camera.py:525]: I think it is better for all outputs to do lazy loading. IE:
```
@property
def isp(self) -> dai.Node.Output:
    if self._isp is None:
        self._create_cam_output(pipeline, self.ISP_STREAM_NAME)
    return self._isp

File: src/robothub/replay_camera.py [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+        self._interleaved = False
+        self._loop = loop
+
+        self._path_type: PathType
+        self._path = self._parse_src(src)
+        if self._path is None:
+            logging.error(f"Error in parsing src: {src}")
+            return
+
+        self.INPUT_CONTROL_STREAM_NAME = "input_control"
```
Comment (PetrNovota) [inline src/robothub/replay_camera.py:114]: I think those constants would be better defined as Enum, because you don't use them as class variables, you have to pass them into class methods as inputs

File: src/robothub/replay_camera.py [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+            return img_frame
+
+        def get_next_frame():
+            if not self._cap.isOpened():
+                return None
+
+            frame = None
+            for _ in range(2):
+                ok, frame = self._cap.read()
+                if not ok and self._loop:
```
Comment (PetrNovota) [inline src/robothub/replay_camera.py:212]: It's a bit unclear to me what this condition does. Maybe something like
```
if not next_frame_exists and self._run_in_loop:
```
is more understandable

File: src/robothub/replay_camera.py [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+
+        time.sleep(self._fps)  # NOTE(miha): Wait for last message to arrive
+        self._cap.release()
+        cv2.destroyAllWindows()
+
+    def start_pooling(self, device: dai.Device):
+        self.thread = threading.Thread(target=self.send_video_frames, args=(device,))
+        self.thread.start()
+
+    def stop_pooling(self):
```
Comment (PetrNovota) [inline src/robothub/replay_camera.py:284]: I don't think this works as expected. the `join` will wait for the thread to terminate, but in the `while True` block the only reason it terminates is when `frame is None`.
maybe I ma missing something, but it looks like this will not terminate the replay, it will either wait for ever in case `self._loop = True` or it will wait for video/image frames to run out

File: src/robothub/replay_camera.py [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+                logging.error(
+                    f"Proccessing time ({process_time}) didn't hit the set camera FPS deadline ({1000.0 / self._fps})"
+                )
+            time_to_sleep = max((1000.0 / self._fps / 1000.0) - process_time, 0)
+            logging.debug(f"process_time: {process_time}, time_to_sleep: {time_to_sleep}")
+            time.sleep(time_to_sleep)
+
+        time.sleep(self._fps)  # NOTE(miha): Wait for last message to arrive
+        self._cap.release()
+        cv2.destroyAllWindows()
```
Comment (PetrNovota) [inline src/robothub/replay_camera.py:278]: why is this line needed?

File: src/robothub/replay_camera.py [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+
+    def _create_cam_output(self, pipeline: dai.Pipeline, stream_name: str, max_data_size: int | None = None):
+        node = pipeline.create(dai.node.XLinkIn)
+        node.setStreamName(stream_name)
+        if max_data_size is not None:
+            node.setMaxDataSize(max_data_size)
+        self._nodes[stream_name] = node
+        return node.out
+
+    def send_video_frames(self, device: dai.Device):
```
Comment (PetrNovota) [inline src/robothub/replay_camera.py:169]: why is this a public method? By the looks of it the expected way to start the replay node is `start_polling`

File: src/robothub/replay_camera.py [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+                    if self._path_type == PathType.VIDEO:
+                        self._cap = cv2.VideoCapture(str(self._path))
+                    else:
+                        self._cap.reset()  # type:ignore
+                    continue
+                break
+
+            return frame
+
+        while True:
```
Comment (PetrNovota) [inline src/robothub/replay_camera.py:223]: this ties to my worry that its not possible to terminate this loop unless it stops on its own.
I think we should use here
```
while rh.app_is_running:
    pass
```
if we want to give the user the option to turn it off manually, then add a second flag

File: src/robothub/replay_camera.py [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+        def get_next_frame():
+            if not self._cap.isOpened():
+                return None
+
+            frame = None
+            for _ in range(2):
+                ok, frame = self._cap.read()
+                if not ok and self._loop:
+                    self._cap.release()
+                    if self._path_type == PathType.VIDEO:
```
Comment (PetrNovota) [inline src/robothub/replay_camera.py:214]: This is a code duplicate from lines 179 - 182, could be a method.

File: src/robothub/replay_camera.py [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+        self._video_width: int = 1920
+        self._video_height: int = 1080
+        self._preview_width: int = 1280
+        self._preview_height: int = 720
+        self._color_order: dai.ColorCameraProperties.ColorOrder = dai.ColorCameraProperties.ColorOrder.BGR
+        self._interleaved = False
+        self._loop = loop
+
+        self._path_type: PathType
+        self._path = self._parse_src(src)
```
Comment (PetrNovota) [inline src/robothub/replay_camera.py:109]: I think we can do a bit better here. The `self._path`variable is only needed to initialize the Video or Image reader into the `self._cap`variable, which is actually not initialized in the constructor. And so the logic where we create the data readers is kind of split around in code. The process is:

- validate src
- initialize `self._cap`

I would vote for doing this in one place

**Review by PetrNovota — CHANGES_REQUESTED** [review]
Good job,
I think its a good start and I hope you don't mind some nit picky comments, but as this is open source I want to have a good quality of code here.

One thing to consider. I always thought it would be nice to have more freedom in configuring the replay node. Lets say I have x videos and I want them to be played in sequence. Or maybe I have videos or files in multiple folders and I want to enter the parent folder + a rule to the file names and the replay would parse recursively all child folders and would replay the files that match the rule.

But lets polish this version and then we can discuss whether its a good time to extend it

File: src/robothub/replay_camera.py (hunk 1 above)
Comment (SleepyPandaBear) [reply src/robothub/replay_camera.py:545]: 1. Lazy loading seems good. 

2. I think the same approach should be used for sending frames to queues (`_self._video_queue.send(...)`), so we can skip computation (`bgr2nv12`, `resize`) if such output is not needed. 

Both 1. and 2. are implemented in new commit.

File: src/robothub/replay_camera.py (hunk 4 above)
Comment (SleepyPandaBear) [reply src/robothub/replay_camera.py]: `stop_pooling` was meant to be called in the `manage_device` function, so you could `join` the thread. It doesn't really make sense to call it in `manage_device`, because the thread would be already terminated. I removed this function.

File: src/robothub/replay_camera.py (hunk 5 above)
Comment (SleepyPandaBear) [reply src/robothub/replay_camera.py]: It is not, the user should handle cv2 window.

File: src/robothub/replay_camera.py (hunk 9 above)
Comment (SleepyPandaBear) [reply src/robothub/replay_camera.py]: I added: `init_cap` and `reset_cap` methods. We init `self._cap` in the constructor and is also responsible for src validation. New methods resolves code duplication that you mentioned in other thread.

**Review by PetrNovota — APPROVED** [review]
